### ISSUE TITLE
vSphere deletes PVs when deleting single node

### ISSUE THREAD

**Issue by horihel** [rancher/rancher#18221]
**What kind of request is this (question/bug/enhancement/feature request):**
Bug Report

**Steps to reproduce (least amount of steps as possible):**
* Create a K8S-Cluster on vSphere (inkl. steps to enable vSphere storage provisioning)
* Create a workload with a volume claim
* Wait for it to be deployed
* Delete the node (in node view) where the pod is running on

**Result:**
* VM gets deleted
* All volumes connected to this VM get deleted as well
* Rancher recreates the VM according to node template
* Rancher/K8S tries to redeploy the workload, but fails, because the volume has been deleted earlier
* PVC needs to be deleted in order to proceed to a working state again
* data in the original volume is lost

**Other details that may be helpful:**
From my limited understanding it looks like Rancher is telling Vsphere to "Delete from disk" the VM, which causes vSphere to delete the VM and all VMDKs currently attached to it. But the volumes created by the provisioner should not be deleted.
This is also a user error, as the node should be drained before deleting it, but in the case of a node not responding, this might not be possible any more.
Maybe Rancher should disconnect all volumes before deleting the VM.

**Environment information**
- Rancher version 2.1.6
- Installation option : single install

**Cluster information**
- Cluster type: Infrastructure Provider
- Machine type: VM and specifications: 7 VMs, 2 CPUs each and 4GB RAM each
- Kubernetes version (use `kubectl version`):

```
Client Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.5", GitCommit:"32ac1c9073b132b8ba18aa830f46b77dcceb0723", GitTreeState:"clean", BuildDate:"2018-06-21T11:46:00Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.6", GitCommit:"b1d75deca493a24a2f87eb1efde1a569e52fc8d9", GitTreeState:"clean", BuildDate:"2018-12-16T04:30:10Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"linux/amd64"}
```

- Docker version (use `docker version`):

```
17.3.2
```

gzrancher/rancher#8983

gzrancher/rancher#11020

gzrancher/rancher#12087

gzrancher/rancher#11675

gzrancher/rancher#12210

**Comment by LinAnt** [comment]
https://github.com/rancher/rancher/issues/15065

This is still an issue.

**Comment by Oats87** [comment]
This is actually an issue with vSphere and not Rancher/Kubernetes, fully.

If you power off a VM in the vSphere UI then "Delete from disk", it will delete all of the attached VMDK's to it. I am not sure what Kubernetes/Rancher could do to prevent this, other than maybe add logic into Rancher to drain the node/ensure detached VMDK's before deleting it.

**Comment by LinAnt** [comment]
https://github.com/kubernetes/kubernetes/issues/75342

**Comment by horihel** [comment]
well on the side of vsphere this is intended/documented behaviour.
Is rancher deleting the VM or is k8s doing that job? 
If rancher is responsible for deleting the VM then it should also make sure that this is a safe operation.

**Comment by skyscooby** [comment]
@Oats87 
Is this going to be fixed? Has your vSphere team at Rancher never been hit with this problem? It seems like a pretty major one that would deserve a bit more attention given it's ramifications on data loss and been untended to in a year...

If you are going to put a fancy +/- button on the UI, with one click and no confirmation or warning.. that implies that a seamless and safe operation is going to occur and Rancher has the responsibility to make it work correctly. Today this is catastrophic in scale down operations.

This is a Rancher issue more than a Vsphere one. If you read the In-Tree Vsphere Cloud docs here (which is what rancher sets up as documented so I understand): https://vmware.github.io/vsphere-storage-for-kubernetes/documentation/high-availability.html#node-vm-failure. It says that the mode of operation on node failure should expects to transition the disks attached to the failed node in a shutdown state.. and is expecting them to remain attached to the failed node. It can some time for the migration to happen but ranchers vsphere orchestration is deleting the node before this is confirmed or verified as being complete and in turn it is triggering vpshere to delete all volumes still attached to it when deleted.. 

A few options to address might include:
- an option in vsphere clusters to simply power off machines on scale down and power them on on scale up if they already exist only creating a new node when it hasn't been created yet..
- for rancher OS provisioned clusters you know how many disks are part of the OS.. you could count the disks attached to the VM through the API and wait to delete it until only the OS disk(s) are left.
- an option to disable the +/- button for vsphere clusters when the cloud provisioner is enabled.

**Comment by skyscooby** [comment]
Perhaps a scan of the VM for any drives attached matching a regex (not sure its correct) like `kubevols/.*pvc.*vmdk` could trigger delaying the delete operation up to a certain time and if after that delay it's still not clean then just power down the VM (don't delete) and display a message that we couldn't confirm it was safe to delete the VM due to potentially stuck PV's.

**Comment by janeczku** [comment]
Looks like one of the two could be implemented to address this:

1) Node should be drained implicitly before deleting the node and VM
2) Delete node, Poweroff VM, wait for cloud provider to detach the volumes, then delete VM

Also, some relevant changes in the out-of-tree CSI provider on vSphere 6.7u3: https://github.com/vmware/vsphere-storage-for-kubernetes/issues/55#issuecomment-536084195

**Comment by daneb255** [comment]
bump ..

**Comment by luthermonson** [comment]
There is a larger body of work that Rancher will be doing to support out of tree providers which will eventually happen so adding supporting for the vsphere out of tree provider is out of scope for this issue. To fix this now we are going to add support for vsphere clusters with the cloud provider to do a proper drain from Rancher while deleting the node which will let the in tree provider do it's work before we call the delete vm in vsphere. this method should work fine when using the `-` button and the delete from the `...` menu.

**Comment by skyscooby** [comment]
@luthermonson awesome news!. Any idea when the fix will be released? Will it backport to the 2.4 release stream as well?

**Comment by luthermonson** [comment]
@skyscooby it's going into 2.4 first and I will forward port to 2.5+

**Comment by bmdepesa** [comment]
reproduced in `rancher/rancher:v2.48`

* Create a vsphere cluster w/ cloud provider, with multiple worker nodes
* Create a VS storage class
* Deploy a workload with a pvc from storage class -> See this scheduled correctly 
* Get yaml output of one of the `node.management.cattle.io` objects from the Rancher management plane
* Delete the node the workload is running on -> Workload becomes unavailable with error:
```
Multi-Attach error for volume "pvc-819d0a49-3b22-4cba-b959-498ed955e2d6" Volume is already exclusively attached to one node and can't be attached to another
```
* Upgrade to `rancher/rancher:v2.4-3974-head`
* From Rancher management plane, get the yaml output of one of the `node.management.cattle.io` objects, and see the finalizer removed, and annotation added:
 ```
 nodes.management.cattle.io/user-node-remove-cleanup: "true"
 ```
* Deployed another workload with pvc from storage class to worker 1
* Delete worker 1
* Worker 1 gets cordoned and drained
* Workload gets scheduled to another worker, volume is not deleted in vCenter, and is attached to the new worker node where workload is scheduled

Leaving open for testing in v2.5/master lines

**Comment by bmdepesa** [comment]
Reopening until merged in v2.5/master

**Comment by bmdepesa** [comment]
`rancher/rancher:master-head 9124f6ed1`
`rancher/rancher:v2.5-head d87f58a`

In each version: 
* Create vSphere cluster w/ cloud provider, create a storage class
* Deployed a workload with pvc from storage class
* Delete the worker node where the workload is scheduled
* The deleted node gets cordoned and drained
* Workload gets scheduled to another worker, volume is not deleted in vCenter, and is attached to the new worker node where workload is scheduled
